qa: Add tests to validate synced images on rbd-mirror

ajarr opened the pull request:
Introduce functional tests to validate that the images under
workloads are correctly mirrored between two clusters using snapshot
based mirroring.

Run workload on a primary image using a krbd or nbd client. Take
mirror snapshots of the image under workload. Unmount the mapped image
and calculate its MD5 checksum before demoting it. After demotion,
wait for the mirror status of the image to be 'up+unknown' in both
the clusters. This is to make sure that the non-primary image in the
other cluster is ready to be promoted. Now promote the non-primary
image in the other cluster. Map the promoted image and calculate its
MD5 checksum. Verify that the checksums of the demoted and promoted
images in the two clusters are the same.

The above test is run as part of two different workunits:
 - a workunit that validates the syncing of multiple mirrored images
   with workloads running on them
 - another workunit that validates the syncing of a single mirrored
   image with workload running on it and the image is set as primary
   alternatively between the two clusters, as it happens during
   failover and failback scenarios.

Fixes: https://tracker.ceph.com/issues/61617
Signed-off-by: Ramana Raja <rraja@redhat.com>
Co-authored-by: Ilya Dryomov <idryomov@redhat.com>
Co-authored-by: Christopher Hoffman <choffman@redhat.com>
<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [x] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

ajarr commented on qa/workunits/rbd/compare_mirror_images.sh:
```diff
@@ -0,0 +1,154 @@
+            | grep non_primary | grep demote | grep -v "%" || true)
+    if [ "$RET" != "" ]; then
+      echo demoted snapshot received, continuing
+      sleep 30s # wait a bit for it to propagate
```
sleeptime was 10s in the original PR https://github.com/ceph/ceph/pull/52132 . Increased it to 30s to check whether it helps with MD5 hashes failing to match for only nbd workloads. Still fails,
 https://pulpito.ceph.com/rraja-2023-12-06_01:23:15-rbd:mirror-main-distro-default-smithi/

ajarr replied:
Got it!

ajarr commented on qa/workunits/rbd/compare_mirror_images.sh:
```diff
@@ -0,0 +1,154 @@
+  demote_image ${CLUSTER1} ${POOL} ${IMG}
+
+  # demote primary image and calculate hash of its latest mirror snapshot
+  DEMOTE=$(rbd --cluster ${CLUSTER1} snap ls --all ${POOL}/${IMG} \
```
better to use local variables in this function

ajarr commented on qa/workunits/rbd/compare_mirror_images.sh:
```diff
@@ -0,0 +1,158 @@
+
+  promote_image ${CLUSTER2} ${POOL} ${img}
+
+  # promote non-primary image and calculate hash of its latest mirror snapshot
```
move this comment above the `promote_image` function call

ajarr commented on qa/workunits/rbd/compare_mirror_images.sh:
```diff
@@ -0,0 +1,158 @@
+      ${POOL}/${img}
+  demote_image ${CLUSTER1} ${POOL} ${img}
+
+  # demote primary image and calculate hash of its latest mirror snapshot
```
move this comment above `demote_image` function call

ajarr commented on qa/workunits/rbd/compare_mirror_images.sh:
```diff
@@ -0,0 +1,158 @@
+
+  for j in {1..10}; do
+    IMG=${IMG_PREFIX}${j}
+    remove_image ${CLUSTER2} ${POOL} ${IMG}
```
to ensure that the removal of the primary/promoted image propagates, maybe we need to wait for the `mirror image status` of the non-primary image to be in `up+replaying` state?
```
wait_for_status_in_pool_dir ${CLUSTER1} ${POOL} ${IMG} 'up+replaying'
// and then...
remove_image ${CLUSTER2} ${POOL} ${IMG}
```

ajarr commented:
jenkins test make check

ajarr commented:
Tests passed,
https://pulpito.ceph.com/rraja-2024-01-17_00:32:33-rbd:mirror-main-distro-default-smithi/

Multiple runs of rbd-mirror-snapshot-workunit-compare-mirror-images-nbd, which used to failed before using`-o try-netlink` option to map images to NBD devices,
https://pulpito.ceph.com/rraja-2024-01-17_00:33:38-rbd:mirror-main-distro-default-smithi/

idryomov commented on qa/workunits/rbd/compare_mirror_image_alternate_primary.sh:
```diff
@@ -0,0 +1,141 @@
+
+  KERNEL_TAR_URL="https://cdn.kernel.org/pub/linux/kernel/v4.x/linux-4.14.280.tar.gz"
+  sudo wget $KERNEL_TAR_URL -O $mountpt/kernel.tar.gz
+  sudo timeout $timeout bash -c "tar xvfz $mountpt/kernel.tar.gz -C $mountpt \
```
I don't know if you have noticed, but "v" in "xvfz" here is incredibly important.  Omitting it would make the entire test bogus (always pass) because `pv` wouldn't have anything to rate limit and untar would complete in a seconds.  This would need a comment or, better, reformulating so that `pv` limits untar directly rather than rate limiting the listing of file names on stdout.

ajarr replied:
Thanks for pointing this out. I missed it. Insted, as you suggested, I rate limited the untar.  Rate limiting the untar as follows, `zcat $mountpt/kernel.tar.gz | pv -L 256K --timer | tar xf - -C $mountpt`, took close to an hour in my local setup. The duration of the workload was similar to `tar xvfz $mountpt/kernel.tar.gz -C $mountpt | pv -L 1k --timer` that took approximately 50 minutes to complete.

idryomov replied:
Did you try `zcat $mountpt/kernel.tar.gz | pv -L 256K --timer | tar xf - -C $mountpt` against the bad build?  Does it still reproduce the failure?

ajarr replied:
Yes. See,
http://qa-proxy.ceph.com/teuthology/rraja-2024-02-08_03:12:35-rbd:mirror-wip-ajarr-61617-revert-corruption-fixes-distro-default-smithi/7551544/teuthology.log
http://qa-proxy.ceph.com/teuthology/rraja-2024-02-08_03:12:35-rbd:mirror-wip-ajarr-61617-revert-corruption-fixes-distro-default-smithi/7551548/teuthology.log
I can see the checksum mistmatch ('Mistmatch iteration at ') failure in both.

idryomov reviewed (commented):
Pretty much all comments on `compare_mirror_image_alternate_primary.sh` apply to `compare_mirror_images.sh` too.  I didn't duplicate them there.

ajarr commented on qa/workunits/rbd/compare_mirror_image_alternate_primary.sh:
```diff
@@ -0,0 +1,141 @@
+  # enable mirroring on newly promoted image in the other cluster
+  BDEV=$(sudo rbd --cluster ${CLUSTER2} device map -t ${RBD_DEVICE_TYPE} \
+           ${POOL}/${IMAGE})
+  enable_mirror ${CLUSTER2} ${POOL} ${IMAGE}
```
Got it!. Removed the no-op code.

ajarr commented on qa/workunits/rbd/compare_mirror_images.sh:
```diff
@@ -0,0 +1,193 @@
+  for s in 1 2 4 8 8 8 8 8 8 8 8 16 16; do
+    img_in_list=$(rbd --cluster $cluster ls -p $pool --format=json \
+                    | jq --arg img $image 'index($img)')
+    if [[ $img_in_list == null ]]; then
```
Agree :)  Done

ajarr commented on qa/workunits/rbd/compare_mirror_images.sh:
```diff
@@ -0,0 +1,193 @@
+  done
+
+  for pid in $pids; do
+    wait "$pid"
```
Maybe just a `wait` that'd wait for the completion for all the background processes would be sufficient.?

idryomov replied:
Just `wait` would ignore exit codes for those background processes:
> Waits for each process identified by an ID, which may be a process ID or a job specification, and reports its termination status. If ID is not given, waits for all currently active child processes, and the return status is zero.

... so, if there is a checksum mismatch, it would make it to the output but wouldn't cause the job to fail.

ajarr replied:
:man_facepalming:

Questions on your original comment,  https://github.com/ceph/ceph/pull/54802#discussion_r1466703267
I think it's okay for the script to terminate when one of the background compare job fails.  So,
``` 
for pid in $pids; do
    wait "$pid"
done
```
should be good enough?

Is there a need to wait for all the background compare jobs, and then terminate if at least one of them failed? In that case, we'd have to `set +e` temporarily in the script to collect the exit statuses of all the jobs.

idryomov replied:
> Is there a need to wait for all the background compare jobs, and then terminate if at least one of them failed?

It would be nice, because you could terminate with a bunch of processes still running on the background otherwise.  Not a big of deal for teuthology, but could be an annoyance when running locally.

> In that case, we'd have to `set +e` temporarily in the script to collect the exit statuses of all the jobs.

Not necessarily.  Something like:
```
ret=0
for pid in $pids; do
    wait "$pid" || ret=$?
done
if ((ret != 0)); then
  echo "compare_demoted_promoted_images failed"
  exit 1
fi
```

ajarr replied:
Thanks for the suggestion. Done.

ajarr commented on qa/workunits/rbd/compare_mirror_image_alternate_primary.sh:
```diff
@@ -0,0 +1,141 @@
+            | jq 'last' \
+            | jq 'select(.name | contains("non_primary"))' \
+            | jq 'select(.namespace.state == "demoted")' \
+            | jq 'select(.namespace.complete == true)')
```
Tested the following change to `wait_for_non_primary_demoted_mirror_snap()`
```
-  local ret
+  local demoted=false

   for s in 1 2 4 8 8 8 8 8 8 8 8 16 16; do
-    ret=$(rbd --cluster $cluster snap ls --all $pool/$image --format=json \
-            | jq 'last' \
-            | jq 'select(.name | contains("non_primary"))' \
-            | jq 'select(.namespace.state == "demoted")' \
-            | jq 'select(.namespace.complete == true)')
-    if [ "$ret" != "" ]; then
+    demoted=$(rbd --cluster $cluster snap ls --all $pool/$image --format=json \
+              | jq '(last | .name | startswith(".mirror.non_primary")) and
+                    (last | .namespace.state == "demoted") and
+                    (last | .namespace.complete == true)')
+
+    if [ "$demoted" = true ]; then
```

On testing, discovered that waiting for non primary demoted snapshot to be synced completely isn't sufficient to check that the image is ready to be promoted.  The test runs failed while trying to promote the image with being unable to create the promotion snapshot.
e.g., http://qa-proxy.ceph.com/teuthology/rraja-2024-01-30_17:01:34-rbd:mirror-main-distro-default-smithi/7539211/teuthology.log
In /a/rraja-2024-01-30_17:01:34-rbd:mirror-main-distro-default-smithi/7539211/remote/smithi175/log/ cluster2-client.mirror.29589.log.gz ,

```
2024-01-30T17:29:55.200+0000 7fd5ea73f640 10 librbd::ImageWatcher: 0x7fd5c8008a00 remote snap_create request: [4281,140556227323456,1] [mirror state=primary, complete=0, mirror_peer_uuids=05b2f9ef-e346-45e0-8d98-4a086cfaea16, clean_since_snap_id=head] .mirror.primary.07a17e35-3bf6-4b6e-bf88-133a68eabc73.61a582ff-e1d8-48f2-b699-b9c250763ba8 2
2024-01-30T17:29:55.200+0000 7fd5ea73f640 20 librbd::ExclusiveLock: 0x7fd5cc004190 accept_request=0 (request_type=0)
2024-01-30T17:29:55.200+0000 7fd5ea73f640 10 librbd::Watcher::C_NotifyAck 0x558063f59e90 finish: r=0
2024-01-30T17:29:55.200+0000 7fd5e9f3e640 20 librbd::watcher::Notifier: 0x7fd5c8008a80 handle_notify: r=0
2024-01-30T17:29:55.200+0000 7fd5e9f3e640 20 librbd::watcher::Notifier: 0x7fd5c8008a80 handle_notify: pending=0
2024-01-30T17:29:55.200+0000 7fd5e9f3e640 20 librbd::image_watcher::NotifyLockOwner: 0x7fd5d0001e00 handle_notify: r=0
2024-01-30T17:29:55.200+0000 7fd5e9f3e640 20 librbd::image_watcher::NotifyLockOwner: 0x7fd5d0001e00 handle_notify client responded with r=-30
2024-01-30T17:29:55.200+0000 7fd5e9f3e640 20 librbd::ImageWatcher: remove_async_request: [4281,140556227323456,1]
2024-01-30T17:29:55.200+0000 7fd5e9f3e640 20 librbd::Operations: handle_remote_request: r=-30
2024-01-30T17:29:55.200+0000 7fd5e9f3e640 20 librbd::ImageState: 0x558063f568a0 handle_update_notification: refresh_seq = 1, last_refresh = 0
2024-01-30T17:29:55.200+0000 7fd5e9f3e640 20 librbd::ImageState: 0x7fd5e00043b0 ImageUpdateWatchers::notify
2024-01-30T17:29:55.200+0000 7fd5e9f3e640 15 librbd::mirror::snapshot::CreatePrimaryRequest: 0x7fd5c8002790 handle_create_snapshot: r=-30
2024-01-30T17:29:55.200+0000 7fd5e9f3e640 -1 librbd::mirror::snapshot::CreatePrimaryRequest: 0x7fd5c8002790 handle_create_snapshot: failed to create mirror snapshot: (30) Read-only file system
2024-01-30T17:29:55.200+0000 7fd5e9f3e640 15 librbd::mirror::snapshot::CreatePrimaryRequest: 0x7fd5c8002790 finish: r=-30
2024-01-30T17:29:55.200+0000 7fd5e9f3e640 15 librbd::mirror::snapshot::PromoteRequest: 0x7fd5c801bc60 handle_create_promote_snapshot: r=-30
2024-01-30T17:29:55.200+0000 7fd5e9f3e640 -1 librbd::mirror::snapshot::PromoteRequest: 0x7fd5c801bc60 handle_create_promote_snapshot: failed to create promote snapshot: (30) Read-only file system
2024-01-30T17:29:55.200+0000 7fd5e9f3e640 15 librbd::mirror::snapshot::PromoteRequest: 0x7fd5c801bc60 finish: r=-30
2024-01-30T17:29:55.200+0000 7fd5e9f3e640 20 librbd::mirror::PromoteRequest: 0x7fd5cc003550 handle_promote: r=-30
2024-01-30T17:29:55.200+0000 7fd5e9f3e640 -1 librbd::mirror::PromoteRequest: 0x7fd5cc003550 handle_promote: failed to promote image: (30) Read-only file system
2024-01-30T17:29:55.200+0000 7fd5e9f3e640 20 librbd::mirror::PromoteRequest: 0x7fd5cc003550 finish: r=-30
2024-01-30T17:29:55.200+0000 7fd5e9f3e640 20 librbd::ImageState: 0x558063f568a0 handle_update_notification: refresh_seq = 2, last_refresh = 0
2024-01-30T17:29:55.200+0000 7fd5e9f3e640 20 librbd::ImageState: 0x7fd5e00043b0 ImageUpdateWatchers::notify
2024-01-30T17:29:55.200+0000 7fd5ed422e00 -1 librbd::api::Mirror: image_promote: failed to promote image
```

Instead of using `wait_for_non_primary_demoted_mirror_snap()` ,  Ilya suggested that we wait for the rbd mirror image status to be `up+unknown`,
```
wait_for_status_in_pool_dir ${CLUSTER1} ${POOL} ${img} 'up+unknown'
wait_for_status_in_pool_dir ${CLUSTER2} ${POOL} ${img} 'up+unknown'
```
to be ready for promoting the image.

ajarr commented on qa/workunits/rbd/compare_mirror_image_alternate_primary.sh:
```diff
@@ -0,0 +1,141 @@
+  KERNEL_TAR_URL="https://cdn.kernel.org/pub/linux/kernel/v4.x/linux-4.14.280.tar.gz"
+  sudo wget $KERNEL_TAR_URL -O $mountpt/kernel.tar.gz
+  sudo timeout $timeout bash -c "tar xvfz $mountpt/kernel.tar.gz -C $mountpt \
+    | pv -L 1k --timer &> /dev/null" || true
```
Got it

ajarr commented on qa/workunits/rbd/compare_mirror_images.sh:
```diff
@@ -0,0 +1,148 @@
+      ${POOL}/${img}
+
+  demote_image ${CLUSTER1} ${POOL} ${img}
+  wait_for_status_in_pool_dir ${CLUSTER1} ${POOL} ${img} 'up+unknown'
```
Fails here for `compare_mirror_images.sh` workloads
See, http://qa-proxy.ceph.com/teuthology/rraja-2024-02-05_20:53:19-rbd:mirror-main-distro-default-smithi/7547714/teuthology.log
<pre>
2024-02-05T21:45:07.311 INFO:tasks.workunit.cluster1.client.mirror.smithi192.stderr:+ rbd --cluster cluster1 mirror pool status mirror --verbose --format xml
2024-02-05T21:45:07.312 INFO:tasks.workunit.cluster1.client.mirror.smithi192.stderr:-:1.1: Document is empty
2024-02-05T21:45:07.313 INFO:tasks.workunit.cluster1.client.mirror.smithi192.stderr:
2024-02-05T21:45:07.313 INFO:tasks.workunit.cluster1.client.mirror.smithi192.stderr:^
2024-02-05T21:45:07.313 INFO:tasks.workunit.cluster1.client.mirror.smithi192.stderr:-:1.1: Document is empty
2024-02-05T21:45:07.313 INFO:tasks.workunit.cluster1.client.mirror.smithi192.stderr:
2024-02-05T21:45:07.313 INFO:tasks.workunit.cluster1.client.mirror.smithi192.stderr:^
2024-02-05T21:45:07.313 INFO:tasks.workunit.cluster1.client.mirror.smithi192.stderr:+ state=
2024-02-05T21:45:07.313 INFO:tasks.workunit.cluster1.client.mirror.smithi192.stderr:+ state=
2024-02-05T21:45:07.313 INFO:tasks.workunit.cluster1.client.mirror.smithi192.stderr:+ echo ''
2024-02-05T21:45:07.313 INFO:tasks.workunit.cluster1.client.mirror.smithi192.stderr:+ grep up+unknown
2024-02-05T21:45:07.313 INFO:tasks.workunit.cluster1.client.mirror.smithi192.stderr:+ echo ''
2024-02-05T21:45:07.313 INFO:tasks.workunit.cluster1.client.mirror.smithi192.stderr:+ grep up+unknown
2024-02-05T21:45:07.314 INFO:tasks.workunit.cluster1.client.mirror.smithi192.stderr:+ test '' '>' '2024-02-05 21:45:06'
2024-02-05T21:45:07.314 INFO:tasks.workunit.cluster1.client.mirror.smithi192.stderr:+ echo ''\''mirror pool status'\'' test failed'
2024-02-05T21:45:07.314 INFO:tasks.workunit.cluster1.client.mirror.smithi192.stderr:'mirror pool status' test failed
2024-02-05T21:45:07.314 INFO:tasks.workunit.cluster1.client.mirror.smithi192.stderr:+ exit 1
2024-02-05T21:45:07.314 INFO:tasks.workunit.cluster1.client.mirror.smithi192.stderr:+ test '' '>' '2024-02-05 21:45:06'
2024-02-05T21:45:07.315 INFO:tasks.workunit.cluster1.client.mirror.smithi192.stderr:+ echo ''\''mirror pool status'\'' test failed'
2024-02-05T21:45:07.315 INFO:tasks.workunit.cluster1.client.mirror.smithi192.stderr:'mirror pool status' test failed
2024-02-05T21:45:07.315 INFO:tasks.workunit.cluster1.client.mirror.smithi192.stderr:+ exit 1
</pre>


However, `compare_mirror_image_alternate_primary` workloads pass. See, https://pulpito.ceph.com/rraja-2024-02-05_20:53:19-rbd:mirror-main-distro-default-smithi/

Need to debug the failure

ajarr replied:
After making the following change to `test_status_in_pool_dir()` the tests passed,
```
diff --git a/qa/workunits/rbd/rbd_mirror_helpers.sh b/qa/workunits/rbd/rbd_mirror_helpers.sh
index f7a3e5de0d9..21c4686a411 100755
--- a/qa/workunits/rbd/rbd_mirror_helpers.sh
+++ b/qa/workunits/rbd/rbd_mirror_helpers.sh
@@ -841,11 +841,8 @@ test_status_in_pool_dir()
 
     local last_update="$(sed -nEe 's/^  last_update: *(.*) *$/\1/p' ${status_log})"
     test_mirror_pool_status_verbose \
-        ${cluster} ${pool} ${image} "${state_pattern}" "${last_update}" &&
-    return 0
+        ${cluster} ${pool} ${image} "${state_pattern}" "${last_update}" || return 1
 
-    echo "'mirror pool status' test failed" >&2
-    exit 1
 }

```
https://github.com/ajarr/ceph/commit/2523ebdb81f1c79c150aa6f8e85685273cc1060e?diff=split&w=0

Tests passed here,
https://pulpito.ceph.com/rraja-2024-02-07_03:00:41-rbd:mirror-main-distro-default-smithi/

It looks like sometimes a mirrored image's status is not captured in `rbd mirror pool status` when there are several images being mirrored. This can explain why without the above change, the workload `compare_mirror_image_alternate_primary.sh`, which involved only one mirrored image, passed.

idryomov replied:
Have you confirmed that it's actually `rbd mirror pool status` that returns no output?  I think it might be a reentrancy issue with the helper: it seems like all `test_mirror_pool_status_verbose()` invocations redirect the output to the same file, which is truncated in the process.  It's possible that the truncate races with one of the later checks.

ajarr replied:
You're right. That's the issue.

How about using a UUID in the names of files that store `mirror image status` or `mirror pool status`?
```
diff --git a/qa/workunits/rbd/rbd_mirror_helpers.sh b/qa/workunits/rbd/rbd_mirror_helpers.sh
index f7a3e5de0d9..ad7427fe97e 100755
--- a/qa/workunits/rbd/rbd_mirror_helpers.sh
+++ b/qa/workunits/rbd/rbd_mirror_helpers.sh
@@ -823,7 +823,7 @@ test_status_in_pool_dir()
     local description_pattern="$5"
     local service_pattern="$6"
 
-    local status_log=${TEMPDIR}/$(mkfname ${cluster}-${pool}-${image}.mirror_status)
+    local status_log=${TEMPDIR}/$(uuidgen).mirror_image_status
     CEPH_ARGS='' rbd --cluster ${cluster} mirror image status ${pool}/${image} |
         tee ${status_log} >&2
     grep "^  state: .*${state_pattern}" ${status_log} || return 1
@@ -856,7 +856,7 @@ test_mirror_pool_status_verbose()
     local state_pattern="$4"
     local prev_last_update="$5"
 
-    local status_log=${TEMPDIR}/$(mkfname ${cluster}-${pool}.mirror_status)
+    local status_log=${TEMPDIR}/$(uuidgen).mirror_pool_status
 
     rbd --cluster ${cluster} mirror pool status ${pool} --verbose --format xml \
         > ${status_log}
```

idryomov replied:
> How about using a UUID in the names of files that store `mirror image status` or `mirror pool status`?

I think something like `mktemp XXXXXXXXXX.mirror_status` would be a bit more natural.

OTOH, I might be missing something, but I don't see a reason to use the file at all.  Why not just capture the output in a variable, potentially echoing it separately so that raw output makes it to the log?

ajarr replied:
Using a variable makes sense. Thanks! Done.

ajarr commented:
And missed to update the commit message

ajarr commented on qa/workunits/rbd/compare_mirror_images.sh:
```diff
@@ -0,0 +1,148 @@
+  local image=$3
+
+  for s in 1 2 4 8 8 8 8 8 8 8 8 16 16; do
+    if [[ -z $(rbd --cluster $cluster ls $pool | grep -w $image) ]]; then
```
Need to add the `-w` option to grep since images have common prefixes, so that pattern `image-primary1` does not match with `image-primary10`.

ajarr commented on qa/workunits/rbd/compare_mirror_image_alternate_primary.sh:
```diff
@@ -0,0 +1,105 @@
+  # rate limit the workload such that the mirror snapshots can be taken as the
+  # contents of the image are progressively changed by the workload.
+  timeout $timeout bash -c "zcat $mountpt/kernel.tar.gz \
+    | pv -L 256K | tar xf - -C $mountpt" || true
```
Thanks for the suggestion!

ajarr commented on qa/workunits/rbd/compare_mirror_images.sh:
```diff
@@ -0,0 +1,148 @@
+
+  for s in 1 2 4 8 8 8 8 8 8 8 8 16 16; do
+    if [[ -z $(rbd --cluster $cluster ls $pool | grep -w $image) ]]; then
+      echo "image:${image} removed from cluster:${cluster} pool:${pool}"
```
Changed to echo only when the image is not removed from the cluster

ajarr commented:
Tests run against `main` passed.

`compare-mirror-images.sh`
https://pulpito.ceph.com/rraja-2024-02-12_20:16:35-rbd:mirror-main-distro-default-smithi/

`compare-mirror-image-alternate-primary.sh'
https://pulpito.ceph.com/rraja-2024-02-12_20:17:23-rbd:mirror-main-distro-default-smithi/

ajarr commented:
Tests running against the branch with data corruption fixes in librbd reverted. Expect to see failed runs only after for tests using nbd clients.

compare-mirror-image-* workloads
https://pulpito.ceph.com/rraja-2024-02-13_03:22:18-rbd:mirror-wip-ajarr-revert-corruption-fixes-feb-12-1-distro-default-smithi/

compare-mirror-image-alternate-primary-* workloads
https://pulpito.ceph.com/rraja-2024-02-13_03:22:37-rbd:mirror-wip-ajarr-revert-corruption-fixes-feb-12-1-distro-default-smithi/

ajarr commented:
jenkins test make check

ajarr commented:
Tests running on main branch with [custom kernel](https://shaman.ceph.com/repos/kernel/wip-revert-exclusive-lock-snapc/a64ccd305b2882c69664fe3c23e858e1968c6dea/) that reverted data corruption fixes in **krbd**.

I expect to see only the krbd workloads failing with MD5 mismatch errors.

compare-mirror-image-{krbd,nbd} workloads
https://pulpito.ceph.com/rraja-2024-02-14_18:23:37-rbd:mirror-main-wip-revert-exclusive-lock-snapc-default-smithi/

compare-mirror-image-alternate-primary-{krbd,nbd} workloads
https://pulpito.ceph.com/rraja-2024-02-14_18:37:40-rbd:mirror-main-wip-revert-exclusive-lock-snapc-default-smithi/


Note: 
Commands used to schedule the runs
```
$ export CEPH_QA_MAIL="rraja@redhat.com"; MACHINE_NAME=smithi; CEPH_BRANCH=main

$ teuthology-suite -v --ceph-repo https://github.com/ceph/ceph-ci.git --suite-repo https://github.com/ajarr/ceph -c $CEPH_BRANCH --suite-branch  wip-61617-feb-12-2024 -m $MACHINE_NAME -s rbd/mirror -p 50 -e $CEPH_QA_MAIL --kernel wip-revert-exclusive-lock-snapc --filter-all compare-mirror-images,ubuntu -l 8

$ teuthology-suite -v --ceph-repo https://github.com/ceph/ceph-ci.git --suite-repo https://github.com/ajarr/ceph -c $CEPH_BRANCH --suite-branch  wip-61617-feb-12-2024 -m $MACHINE_NAME -s rbd/mirror -p 50 -e $CEPH_QA_MAIL --kernel wip-revert-exclusive-lock-snapc --filter-all compare-mirror-image-alternate-primary,ubuntu -l 8
```

ajarr commented:
Note to the reviewer: 
- updated the commit attributions in the latest push; no code changes
- See https://github.com/ceph/ceph/compare/1d33fadb04e6b324b918e0d5c23d31d9190be07d..a5e1b333b6e4ac1e28ed61f37444949bd6a0c115  for code changes since last review

ajarr commented:
jenkins test make check

idryomov commented:
jenkins test api

idryomov commented on qa/workunits/rbd/rbd_mirror_helpers.sh:
```diff
@@ -823,23 +823,23 @@ test_status_in_pool_dir()
-    grep "^  state: .*${state_pattern}" ${status_log} || return 1
-    grep "^  description: .*${description_pattern}" ${status_log} || return 1
+    local status=$(CEPH_ARGS='' rbd --cluster ${cluster} mirror image status \
+                     ${pool}/${image})
```
It looks the output of the command was making it to the log before.  Should it be echoed before running any of the greps?

idryomov commented on qa/workunits/rbd/rbd_mirror_helpers.sh:
```diff
@@ -823,23 +823,23 @@ test_status_in_pool_dir()
-    grep "^  description: .*${description_pattern}" ${status_log} || return 1
+    local status=$(CEPH_ARGS='' rbd --cluster ${cluster} mirror image status \
+                     ${pool}/${image})
+    echo "$status" | grep "^  state: .*${state_pattern}"  || return 1
```
A pipe might be considered excessive for this:
```suggestion
    grep "^  state: .*${state_pattern}" <<< "$status" || return 1
```

idryomov commented on qa/workunits/rbd/compare_mirror_images.sh:
```diff
@@ -0,0 +1,172 @@
+  DEVS=()
+  SNAP_PIDS=''
+  COMPARE_PIDS=''
+  WORKLOAD_PIDS=''
```
Curious why an array is used for devices but not for PIDs?

ajarr replied:
No particular reason. Converted all of them to arrays.

idryomov commented on qa/workunits/rbd/compare_mirror_images.sh:
```diff
@@ -0,0 +1,172 @@
+  fi
+
+  for j in {1..10}; do
+    compare_demoted_promoted_mirror_images ${DEVS[$j-1]} $j &
```
The first argument appears to be redundant -- it can be constructed from `$j` in the function same as `img` and `mntpt`.

ajarr commented on qa/workunits/rbd/compare_mirror_image_alternate_primary.sh:
```diff
@@ -0,0 +1,106 @@
+  sudo chown $(whoami) ${MOUNT}
+  take_mirror_snapshots ${CLUSTER1} ${POOL} ${IMAGE} &
+  slow_untar_workload ${MOUNT}
+  wait $!
```
Good point!

ajarr commented on qa/workunits/rbd/rbd_mirror_helpers.sh:
```diff
@@ -823,23 +823,23 @@ test_status_in_pool_dir()
-        tee ${status_log} >&2
-    grep "^  state: .*${state_pattern}" ${status_log} || return 1
-    grep "^  description: .*${description_pattern}" ${status_log} || return 1
+    local status=$(CEPH_ARGS='' rbd --cluster ${cluster} mirror image status \
```
> * `local status` needs to be declared separately, otherwise potential errors from the command would be swallowed

Got it. Done.

> * not following why `CEPH_ARGS=''` is needed here but not on `rbd mirror pool status` command, suggest just getting rid of it

Looks like it's needed. It "unsets" the `CEPH_ARGS` set by rbd_mirror_helpers.sh.  Otherwise I see that `mirror image status` command does not show the `service` field in its output, and the test fails.
```
2024-02-20T19:53:10.894 INFO:tasks.workunit.cluster1.client.mirror.smithi157.stderr:++ MIRROR_POOL_MODE=image
2024-02-20T19:53:10.894 INFO:tasks.workunit.cluster1.client.mirror.smithi157.stderr:++ export 'CEPH_ARGS=--id mirror'

...

2024-02-20T20:01:04.718 INFO:tasks.workunit.cluster1.client.mirror.smithi157.stderr:+ local status
2024-02-20T20:01:04.719 INFO:tasks.workunit.cluster1.client.mirror.smithi157.stderr:++ rbd --cluster cluster1 mirror image status mirror/image-primary9
2024-02-20T20:01:04.995 INFO:tasks.workunit.cluster1.client.mirror.smithi157.stderr:rbd: failed to query services: (13) Permission denied
2024-02-20T20:01:05.006 INFO:tasks.workunit.cluster1.client.mirror.smithi157.stderr:+ status='image-primary9:
2024-02-20T20:01:05.006 INFO:tasks.workunit.cluster1.client.mirror.smithi157.stderr:  global_id:   04ac2e8f-f9b3-4d81-a169-a4fb5de4143a
2024-02-20T20:01:05.006 INFO:tasks.workunit.cluster1.client.mirror.smithi157.stderr:  state:       up+unknown
2024-02-20T20:01:05.006 INFO:tasks.workunit.cluster1.client.mirror.smithi157.stderr:  description: remote image is not primary
2024-02-20T20:01:05.006 INFO:tasks.workunit.cluster1.client.mirror.smithi157.stderr:  last_update: 2024-02-20 20:01:03
2024-02-20T20:01:05.007 INFO:tasks.workunit.cluster1.client.mirror.smithi157.stderr:  peer_sites:
2024-02-20T20:01:05.007 INFO:tasks.workunit.cluster1.client.mirror.smithi157.stderr:    name: cluster2
2024-02-20T20:01:05.007 INFO:tasks.workunit.cluster1.client.mirror.smithi157.stderr:    state: up+unknown
2024-02-20T20:01:05.007 INFO:tasks.workunit.cluster1.client.mirror.smithi157.stderr:    description: remote image is not primary
2024-02-20T20:01:05.007 INFO:tasks.workunit.cluster1.client.mirror.smithi157.stderr:    last_update: 2024-02-20 20:01:03'
2024-02-20T20:01:05.007 INFO:tasks.workunit.cluster1.client.mirror.smithi157.stderr:+ echo image-primary9: global_id: 04ac2e8f-f9b3-4d81-a169-a4fb5de4143a state: up+unknown description: remote image is not primary last_update: 2024-02-20 20:01:03 peer_sites: name: cluster2 state: up+unknown description: remote image is not primary last_update: 2024-02-20 20:01:03
2024-02-20T20:01:05.008 INFO:tasks.workunit.cluster1.client.mirror.smithi157.stdout:image-primary9: global_id: 04ac2e8f-f9b3-4d81-a169-a4fb5de4143a state: up+unknown description: remote image is not primary last_update: 2024-02-20 20:01:03 peer_sites: name: cluster2 state: up+unknown description: remote image is not primary last_update: 2024-02-20 20:01:03
2024-02-20T20:01:05.008 INFO:tasks.workunit.cluster1.client.mirror.smithi157.stderr:+ grep '^  state: .*up+unknown'
2024-02-20T20:01:05.008 INFO:tasks.workunit.cluster1.client.mirror.smithi157.stdout:  state:       up+unknown
2024-02-20T20:01:05.009 INFO:tasks.workunit.cluster1.client.mirror.smithi157.stderr:+ grep '^  description: .*'
2024-02-20T20:01:05.009 INFO:tasks.workunit.cluster1.client.mirror.smithi157.stdout:  description: remote image is not primary
2024-02-20T20:01:05.009 INFO:tasks.workunit.cluster1.client.mirror.smithi157.stderr:+ '[' -n '' ']'
2024-02-20T20:01:05.010 INFO:tasks.workunit.cluster1.client.mirror.smithi157.stderr:+ echo up+unknown
2024-02-20T20:01:05.010 INFO:tasks.workunit.cluster1.client.mirror.smithi157.stderr:+ grep '^up+'
2024-02-20T20:01:05.011 INFO:tasks.workunit.cluster1.client.mirror.smithi157.stdout:up+unknown
2024-02-20T20:01:05.011 INFO:tasks.workunit.cluster1.client.mirror.smithi157.stderr:+ grep 'service: *mirror..* on '
2024-02-20T20:01:05.012 INFO:tasks.workunit.cluster1.client.mirror.smithi157.stderr:+ return 1
2024-02-20T20:01:05.013 INFO:tasks.workunit.cluster1.client.mirror.smithi157.stderr:+ return 1
2024-02-20T20:01:05.013 INFO:tasks.workunit.cluster1.client.mirror.smithi157.stderr:+ RET=1
2024-02-20T20:01:05.013 INFO:tasks.workunit.cluster1.client.mirror.smithi157.stderr:+ for pid in ${COMPARE_PIDS[@]}
2024-02-20T20:01:05.014 INFO:tasks.workunit.cluster1.client.mirror.smithi157.stderr:+ wait 36290
2024-02-20T20:01:05.014 INFO:tasks.workunit.cluster1.client.mirror.smithi157.stderr:+ RET=1
2024-02-20T20:01:05.014 INFO:tasks.workunit.cluster1.client.mirror.smithi157.stderr:+ (( RET != 0 ))
2024-02-20T20:01:05.014 INFO:tasks.workunit.cluster1.client.mirror.smithi157.stderr:+ echo 'compare_demoted_promoted_image failed'
2024-02-20T20:01:05.014 INFO:tasks.workunit.cluster1.client.mirror.smithi157.stdout:compare_demoted_promoted_image failed
2024-02-20T20:01:05.016 INFO:tasks.workunit.cluster1.client.mirror.smithi157.stderr:+ exit 1
```

Snippet from http://qa-proxy.ceph.com/teuthology/rraja-2024-02-20_19:29:47-rbd:mirror-main-distro-default-smithi/7568961/teuthology.log

ajarr commented:
Tested the latest PR on 'main' branch
https://pulpito.ceph.com/rraja-2024-02-20_21:55:51-rbd:mirror-main-distro-default-smithi/
https://pulpito.ceph.com/rraja-2024-02-21_00:59:14-rbd:mirror-main-distro-default-smithi/

ajarr commented:
Test runs on the latest patchset on 'main' branch
https://pulpito.ceph.com/rraja-2024-02-21_18:33:58-rbd:mirror-main-distro-default-smithi/
https://pulpito.ceph.com/rraja-2024-02-21_19:07:11-rbd:mirror-main-distro-default-smithi/

Since last patchset,
* removed setting CEPH_ARGS='' in the two scripts as the `rbd_mirror_helpers.sh` script sets it to `--id mirror` 
* `mirror pool status`  like `mirror image status` also hit "failed to query services: (13) Permission denied" . So executed the command with CEPH_ARGS=''  and I no longer see  "failed to query services: (13) Permission denied" .

ajarr commented:
Test runs on the latest patchset on 'main' branch
https://pulpito.ceph.com/rraja-2024-02-22_16:50:37-rbd:mirror-main-distro-default-smithi/
https://pulpito.ceph.com/rraja-2024-02-22_16:51:07-rbd:mirror-main-distro-default-smithi/

ajarr commented on qa/workunits/rbd/compare_mirror_images.sh:
```diff
@@ -0,0 +1,170 @@
+    | pv -L 256K | tar xf - -C $mountpt" || ret=$?
+  if ((ret != 124)); then
+    echo "Workload completed prematurely"
+    exit 1
```
Sorry about missing this

ajarr commented on qa/workunits/rbd/rbd_mirror_helpers.sh:
```diff
@@ -856,16 +857,16 @@ test_mirror_pool_status_verbose()
+    local status
+    status=$(CEPH_ARGS='' rbd --cluster ${cluster} mirror pool status ${pool} \
+                 --verbose --format xml)
+    echo $status
```
No problem

idryomov commented:
LGTM!  Could you please schedule 10 jobs of each type against bad builds, just in case?

ajarr commented:
More runs on 'main':
https://pulpito.ceph.com/rraja-2024-02-22_21:27:37-rbd:mirror-main-distro-default-smithi/
https://pulpito.ceph.com/rraja-2024-02-22_21:28:48-rbd:mirror-main-distro-default-smithi/

idryomov commented:
> Runs on 'main' using kernel build (wip-revert-exclusive-lock-snapc), that has the data corruption fixes reverted.

Please schedule the same for a librbd build with reverts for completeness.

ajarr commented:
> > Runs on 'main' using kernel build (wip-revert-exclusive-lock-snapc), that has the data corruption fixes reverted.
> 
> Please schedule the same for a librbd build with reverts for completeness.

Yes, I did try to do it with old build. It looks like I need a new build as some packages are missing. I've initiated the new build.

idryomov commented:
> It looks like I need a new build as some packages are missing. I've initiated the new build.

Yup, IIRC builds expire after two weeks.

ajarr commented:
Had some trouble trying to make the new builds with librbd corruption fixes reverted. In the following runs you should see that the nbd (librbd) jobs failed.
https://pulpito.ceph.com/rraja-2024-02-23_14:33:05-rbd:mirror-wip-ajarr-revert-librbd-corruption-fixes-feb-22-1-distro-default-smithi/
https://pulpito.ceph.com/rraja-2024-02-23_14:33:31-rbd:mirror-wip-ajarr-revert-librbd-corruption-fixes-feb-22-1-distro-default-smithi/

@idryomov let me know what if you need any other testing done
